lib/optimizely/custom_attribute_condition_evaluator.rb

```diff
@@ -0,0 +1,166 @@
+#    limitations under the License.
+#
+module Optimizely
+  class CustomAttributeConditionEvaluator
+    CUSTOM_ATTRIBUTE_CONDITION_TYPE = 'custom_attribute'
+
+    # Conditional match types
+    EXACT_MATCH_TYPE = 'exact'
+    EXISTS_MATCH_TYPE = 'exists'
+    GREATER_THAN_MATCH_TYPE = 'gt'
+    LESS_THAN_MATCH_TYPE = 'lt'
+    SUBSTRING_MATCH_TYPE = 'substring'
+
+    EVALUATORS_BY_MATCH_TYPE = {
+      EXACT_MATCH_TYPE => :exact_evaluator,
+      EXISTS_MATCH_TYPE => :exists_evaluator,
+      GREATER_THAN_MATCH_TYPE => :greater_than_evaluator,
+      LESS_THAN_MATCH_TYPE => :less_than_evaluator,
+      SUBSTRING_MATCH_TYPE => :substring_evaluator
+    }.freeze
+
+    attr_reader :user_attributes
+
+    def initialize(user_attributes)
+      @user_attributes = user_attributes
+    end
+
+    def evaluate(leaf_condition)
+      # Top level method to evaluate audience conditions.
+      #
+      # conditions - Nested array of and/or conditions.
+      #              Example: ['and', operand_1, ['or', operand_2, operand_3]]
+      #
+      # Returns boolean if the given user attributes match/don't match the given conditions,
+      #         nil if the given conditions can't be evaluated.
+
+      return nil unless leaf_condition['type'] == CUSTOM_ATTRIBUTE_CONDITION_TYPE
+
+      condition_match = leaf_condition['match']
+
+      return nil if !condition_match.nil? && !EVALUATORS_BY_MATCH_TYPE.include?(condition_match)
+
+      condition_match = EXACT_MATCH_TYPE if condition_match.nil?
+
+      send(EVALUATORS_BY_MATCH_TYPE[condition_match], leaf_condition)
+    end
+
+    def exact_evaluator(condition)
+      # Evaluate the given exact match condition for the given user attributes.
+      #
+      # Returns boolean true if the user attribute value is equal (===) to the condition value,
+      #                 false if the user attribute value is not equal (!==) to the condition value,
+      #                 nil if the condition value or user attribute value has an invalid type,
+      #                 or if there is a mismatch between the user attribute type and the condition value type.
+
+      condition_value = condition['value']
+      condition_type = condition['value'].class
+
+      user_provided_value = @user_attributes[condition['name']]
+      user_provided_type = @user_attributes[condition['name']].class
+
+      return nil if !value_valid_for_exact_conditions?(user_provided_value) ||
+                    !value_valid_for_exact_conditions?(condition_value) ||
+                    different_types?(condition_type, user_provided_type)
+
+      condition_value == user_provided_value
+    end
+
+    def exists_evaluator(condition)
+      # Evaluate the given exists match condition for the given user attributes.
+      # Returns boolean true if both:
+      #                    1) the user attributes have a value for the given condition, and
+      #                    2) the user attribute value is neither null nor undefined
```

Comment:
Remove undefined. undefined is a type in JS, not in ruby

Reply:
undefined is correct and updated null to nil.